Framework/Catalog/inc/MantidCatalog/ONCat.h

```diff
@@ -0,0 +1,158 @@
+  an authenticated user to have their tokens invalidated  *eventually*.
+  At that point, any call to the API will fail, and an error will be written
+  to the log asking them to login again.
+
+  @author Peter Parker
+  @date 2018
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_CATALOG_DLL ONCat {
+public:
+  static ONCat fromMantidSettings();
+
+  ONCat() = delete;
+  ONCat(const ONCat &other);
+  ~ONCat();
+
+  bool isUserLoggedIn() const;
+
+  void login(const std::string &username, const std::string &password);
+  void logout();
+
+  ONCatEntity retrieve(const std::string &resourceNamespace,
+                       const std::string &resource,
+                       const std::string &identifier,
+                       const QueryParameters &queryParameters);
+  std::vector<ONCatEntity> list(const std::string &resourceNamespace,
+                                const std::string &resource,
+                                const QueryParameters &queryParameters);
+
+  //////////////////////////////////////////////////////////////////////
+  // Exposed publicly for testing purposes only.
+  //////////////////////////////////////////////////////////////////////
+  ONCat(const std::string &url, IOAuthTokenStore_uptr tokenStore,
+        OAuthFlow flow, const std::string &clientId,
+        const boost::optional<std::string> &clientSecret = boost::none);
+  void refreshTokenIfNeeded();
+  void refreshTokenIfNeeded(const DateAndTime &currentTime);
+  void setInternetHelper(
+      std::shared_ptr<Mantid::Kernel::InternetHelper> internetHelper);
+  //////////////////////////////////////////////////////////////////////
+
+private:
+  void sendAPIRequest(const std::string &uri,
+                      const QueryParameters &queryParameters,
+                      std::ostream &response);
+
+  std::string m_url;
+  IOAuthTokenStore_sptr m_tokenStore;
+  std::string m_clientId;
+  boost::optional<std::string> m_clientSecret;
+
+  OAuthFlow m_flow;
+  std::shared_ptr<Mantid::Kernel::InternetHelper> m_internetHelper;
```

Comment:
I think this should be a `std::unique_ptr` created using (so it is portable) `Mantid::Kernel::make_unique`.

Reply:
So, to clarify, you would prefer access to a `unique_ptr` version of `m_internetHelper` (required so that we can make assertions on mocked "internet helpers" in the tests) to be granted via a raw, "observing" pointer?

```c++
Mantid::Kernel::InternetHelper * ONCat::internetHelper() {
  return m_internetHelper.get();
}
```

Something like that?

I'll admit I discounted this approach without much thought, mainly out of an indiscriminate, blanket aversion to raw pointers.  I've done [some reading](https://stackoverflow.com/a/15649077/778572), though, and I _think_ I've convinced myself that this is the recommended approach.  (At least until [`std::observer_ptr`](https://en.cppreference.com/w/cpp/experimental/observer_ptr) comes along?)

Reply:
I hadn't thought that deeply on it. I just noticed that it was an object that only one class needed access to so `std::shared_ptr` wasn't the right concept. The thought behind the suggestion was that you'd get the lifecycle handled for "free" since the pointer would be cleaned up automatically when an ONCat object was deleted.

Reply:
I'll go with the change I outlined above and see how we get on.

Reply:
Actually Pete, scratch that.  I started making the change and I think a shared pointer here is just cleaner, mainly because of what's happening in the tests.

With a `unique_ptr` approach, when we construct mock `MockONCatAPI` objects (children of `Mantid::Kernel::InternetHelper`), we would be `move`-ing them into the `ONCat` class, but then when we needed them again we would have to retrieve as an observing pointer, and then cast back to `MockONCatAPI` objects before being able to call, for example, `MockONCatAPI::allResponsesCalledOnce`.

A `shared_ptr` approach means we could just keep ahold of the `MockONCatAPI` objects in the test scope, pass them in, and then call them directly after we're finished using whatever functionality is under test.

Not sure how much all that made sense, but unless you disagree I'll keep the `shared_ptr` approach and just make the [change below](https://github.com/mantidproject/mantid/pull/22635#discussion_r223466963).